--- phantomjs-1.9.0/build.sh	2012-09-22 21:37:28.000000000 +0100
+++ phantomjs-1.9.0/build.sh	2012-11-18 04:47:00.588873350 +0000
@@ -43,9 +43,9 @@
             shift
             QMAKE_ARGS=$1
             shift;;
-        "--jobs")
+        "--makeopts")
             shift
-            COMPILE_JOBS=$1
+            MAKEOPTS=$1
             shift;;
         "--confirm")
             BUILD_CONFIRM=1
@@ -66,29 +66,6 @@
 done
 
 
-if [[ "$BUILD_CONFIRM" -eq 0 ]]; then
-cat << EOF
-----------------------------------------
-               WARNING
-----------------------------------------
-
-Building PhantomJS from source takes a very long time, anywhere from 30 minutes
-to several hours (depending on the machine configuration). It is recommended to
-use the premade binary packages on supported operating systems.
-
-For details, please go the the web site: http://phantomjs.org/download.html.
-
-EOF
-
-    echo "Do you want to continue (y/n)?"
-    read continue
-    if [[ "$continue" != "y" ]]; then
-        exit 1
-    fi
-    echo
-    echo
-fi
-
-cd src/qt && ./preconfig.sh --jobs $COMPILE_JOBS --qt-config "$QT_CFG" && cd ../..
-src/qt/bin/qmake $QMAKE_ARGS
-make -j$COMPILE_JOBS
+cd src/qt && ./preconfig.sh --makeopts $MAKEOPTS
+cd ../.. && src/qt/bin/qmake
+make $MAKEOPTS
 
