--- rsyslog-8.4.0/runtime/msg.c	2014-08-18 06:14:57.000000000 -0400
+++ rsyslog-8.4.0-new/runtime/msg.c	2014-09-28 09:58:05.460532569 -0400
@@ -4074,7 +4074,11 @@
 
 			err = tokener->err;
 			if(err != json_tokener_continue)
+#if JSON_C_MINOR_VERSION >= 12
+				errMsg = json_tokener_error_desc(err);
+#else
 				errMsg = json_tokener_errors[err];
+#endif
 			else
 				errMsg = "Unterminated input";
 		} else if(!json_object_is_type(json, json_type_object))
