--- rrdtool-1.4.8-old/src/rrd_tool.c	2014-03-14 16:59:40.260043376 -0400
+++ rrdtool-1.4.8/src/rrd_tool.c	2014-03-14 17:11:48.720944511 -0400
@@ -737,7 +737,7 @@
                 printf("  <%s>\n", META_TAG);
             }
             else {
-                printf("{ about: 'RRDtool xport JSON output',\n  meta: {\n");
+                printf("{ \"about\": \"RRDtool xport JSON output\",\n  \"meta\": {\n");
             }
 
 
@@ -766,7 +766,7 @@
                 char     *entry = NULL;
                 entry = legend_v[j];
                 if (json){
-                    printf("      '%s'", entry);
+                    printf("      \"%s\"", entry);
                     if (j < col_cnt -1){
                         printf(",");
                     }
